src/module.c

```diff
@@ -4288,6 +4299,7 @@ int RM_ListInsert(RedisModuleKey *key, long index, RedisModuleString *value) {
         /* Insert before the first element => push head. */
         return RM_ListPush(key, REDISMODULE_LIST_HEAD, value);
     }
+    listTypeTryConvertListpack(key->value, &value, 0, 0, moduleFreeListIterator, key);
```

Comment:
So far, the module implementation only used the high level abstraction of the list type. I think the abstraction is better if it doesn't need to know about the internals, such as encodings.

Is it possible to let listTypeInsert handle conversion from listpack to quicklist?

Reply:
Ah, it's dangerous....

What if we just name the functions so it doesn't expose the internals such as encoding, but rather the purpose from the caller's perspecive? For example, instead of `listTypeTryConvertListpack` it can be called `listTypeMakeSpaceForValues` or something like that?

Similarily, converting from quicklist to listpack can be called something like `listTypeShrink`. WDYT?

Reply:
@zuiderkwast Good idea.
I prefer `listTypeTryConversion` and `listTypeTryConverionForValues` more, which more clearly indicate that the encoding will be converted.

Reply:
Sunds good (although I think indicating that the encoding will be converted can be something internal to the list type that the caller doesn't need to know).

Reply:
In the end I chose `listTypeTryConversion` and `listTypeTryConverionForValues`, at least they are more consistent with other conversion function like `zset` or `hash`.